src/Neo/Network/P2P/Payloads/Signer.cs

```diff
@@ -190,16 +189,30 @@ void IInteroperable.FromStackItem(VM.Types.StackItem stackItem)
             throw new NotSupportedException();
         }
 
-        VM.Types.StackItem IInteroperable.ToStackItem(ReferenceCounter referenceCounter)
+        VM.Types.StackItem IInteroperable.ToStackItem(ApplicationEngine applicationEngine)
         {
-            return new VM.Types.Array(referenceCounter, new VM.Types.StackItem[]
+            if (!applicationEngine.IsHardforkEnabled(Hardfork.HF_Cockatrice))
             {
+                // It will FAULT if any of the values are null
+
+                return new VM.Types.Array(applicationEngine.ReferenceCounter,
+                [
+                    Account.ToArray(),
+                    (byte)Scopes,
+                    new VM.Types.Array(applicationEngine.ReferenceCounter, AllowedContracts.Select(u => new VM.Types.ByteString(u.ToArray()))),
+                    new VM.Types.Array(applicationEngine.ReferenceCounter, AllowedGroups.Select(u => new VM.Types.ByteString(u.ToArray()))),
+                    new VM.Types.Array(applicationEngine.ReferenceCounter, Rules.Select(u => u.ToStackItem(applicationEngine)))
+                ]);
+            }
+
+            return new VM.Types.Array(applicationEngine.ReferenceCounter,
+            [
                 Account.ToArray(),
                 (byte)Scopes,
-                new VM.Types.Array(referenceCounter, AllowedContracts.Select(u => new VM.Types.ByteString(u.ToArray()))),
-                new VM.Types.Array(referenceCounter, AllowedGroups.Select(u => new VM.Types.ByteString(u.ToArray()))),
-                new VM.Types.Array(referenceCounter, Rules.Select(u => u.ToStackItem(referenceCounter)))
-            });
+                Scopes.HasFlag(WitnessScope.CustomContracts) ? new VM.Types.Array(applicationEngine.ReferenceCounter, AllowedContracts.Select(u => new VM.Types.ByteString(u.ToArray()))) : VM.Types.StackItem.Null,
+                Scopes.HasFlag(WitnessScope.CustomGroups) ? new VM.Types.Array(applicationEngine.ReferenceCounter, AllowedGroups.Select(u => new VM.Types.ByteString(u.ToArray()))) : VM.Types.StackItem.Null,
+                Scopes.HasFlag(WitnessScope.WitnessRules) ? new VM.Types.Array(applicationEngine.ReferenceCounter, Rules.Select(u => u.ToStackItem(applicationEngine))) : VM.Types.StackItem.Null
```

Comment:
Which is the way it's handled here: https://github.com/nspcc-dev/neo-go/blob/bc158eed519cd7be5a41c507d6057ac1ee872e5a/pkg/core/transaction/signer.go#L82-L84

Reply:
But it will fix the problem, it's possible to do a special transaction for use this "exception" change

Reply:
How?

I've created this contract (Go):
```
package test

import (
        "github.com/nspcc-dev/neo-go/pkg/interop/runtime"
)

// Test does the test.
func Test() any {
        return runtime.CurrentSigners()
}
```
with this config (yeah, we need it for NeoGo):
```
name: "CurrentSigners example"
sourceurl: https://github.com/nspcc-dev/neo-go/
supportedstandards: []
```
It compiles down to the following code:
```
INDEX    OPCODE     PARAMETER
0        SYSCALL    System.Runtime.CurrentSigners (acf1188b)    <<
5        RET        
```
with a rather simple manifest:
```
{
   "trusts" : [],
   "extra" : null,
   "abi" : {
      "methods" : [
         {
            "offset" : 0,
            "safe" : false,
            "name" : "test",
            "parameters" : [],
            "returntype" : "Any"
         }
      ],
      "events" : []
   },
   "name" : "CurrentSigners example",
   "permissions" : [],
   "supportedstandards" : [],
   "features" : {},
   "groups" : []
}
```
This contract is deployed to testnet, [b8c580fed12519542b294cf487c480abead705de](https://dora.coz.io/contract/neo3/testnet/0xb8c580fed12519542b294cf487c480abead705de). I've tried test invocations:
```
$ ./bin/neo-go contract testinvokefunction -r https://rpc.t5.n3.nspcc.ru:20331 b8c580fed12519542b294cf487c480abead705de test
{
  "state": "HALT",
  "gasconsumed": "98451",
  "script": "wh8MBHRlc3QMFN4F1+qrgMSH9EwpK1QZJdH+gMW4QWJ9W1I=",
  "stack": [
    {
      "type": "Array",
      "value": [
        {
          "type": "Array",
          "value": [
            {
              "type": "ByteString",
              "value": "AAAAAAAAAAAAAAAAAAAAAAAAAAA="
            },
            {
              "type": "Integer",
              "value": "0"
            },
            {
              "type": "Array",
              "value": []
            },
            {
              "type": "Array",
              "value": []
            },
            {
              "type": "Array",
              "value": []
            }
          ]
        }
      ]
    }
  ],
  "exception": null,
  "notifications": []
}
$ ./bin/neo-go contract testinvokefunction -r https://rpc.t5.n3.nspcc.ru:20331 b8c580fed12519542b294cf487c480abead705de test -- NXN73aeWgm8DLebwZXaKaFtZXmDiMHGmKD:CalledByEntry
{
  "state": "HALT",
  "gasconsumed": "98451",
  "script": "wh8MBHRlc3QMFN4F1+qrgMSH9EwpK1QZJdH+gMW4QWJ9W1I=",
  "stack": [
    {
      "type": "Array",
      "value": [
        {
          "type": "Array",
          "value": [
            {
              "type": "ByteString",
              "value": "fZrCUAp6XHd0agwtIy6wf49ymNw="
            },
            {
              "type": "Integer",
              "value": "1"
            },
            {
              "type": "Array",
              "value": []
            },
            {
              "type": "Array",
              "value": []
            },
            {
              "type": "Array",
              "value": []
            }
          ]
        }
      ]
    }
  ],
  "exception": null,
  "notifications": []
}
$ ./bin/neo-go contract testinvokefunction -r http://seed5t5.neo.org:20332 b8c580fed12519542b294cf487c480abead705de test
{
  "state": "HALT",
  "gasconsumed": "98451",
  "script": "wh8MBHRlc3QMFN4F1+qrgMSH9EwpK1QZJdH+gMW4QWJ9W1I=",
  "stack": [
    {
      "type": "Any"
    }
  ],
  "exception": null,
  "notifications": []
}
$ ./bin/neo-go contract testinvokefunction -r http://seed5t5.neo.org:20332 b8c580fed12519542b294cf487c480abead705de test -- NXN73aeWgm8DLebwZXaKaFtZXmDiMHGmKD:CalledByEntry
{
  "state": "HALT",
  "gasconsumed": "98451",
  "script": "wh8MBHRlc3QMFN4F1+qrgMSH9EwpK1QZJdH+gMW4QWJ9W1I=",
  "stack": [
    {
      "type": "Array",
      "value": [
        {
          "type": "Array",
          "value": [
            {
              "type": "ByteString",
              "value": "fZrCUAp6XHd0agwtIy6wf49ymNw="
            },
            {
              "type": "Integer",
              "value": "1"
            },
            {
              "type": "Array",
              "value": []
            },
            {
              "type": "Array",
              "value": []
            },
            {
              "type": "Array",
              "value": []
            }
          ]
        }
      ]
    }
  ],
  "exception": null,
  "notifications": []
}
```
And I've tried a real one in [ed1d5081dcef12d2c437ae8566d8040cf0a475e7e354c9acbd0f674a372106f8](https://dora.coz.io/transaction/neo3/testnet/0xed1d5081dcef12d2c437ae8566d8040cf0a475e7e354c9acbd0f674a372106f8):
```
$ curl -sd '{ "jsonrpc": "2.0", "id": 1, "method": "getapplicationlog", "params": ["ed1d5081dcef12d2c437ae8566d8040cf0a475e7e354c9acbd0f674a372106f8"] }' http://seed5t5.neo.org:20332 | json_pp
{
   "result" : {
      "executions" : [
         {
            "trigger" : "Application",
            "vmstate" : "HALT",
            "exception" : null,
            "gasconsumed" : "98451",
            "notifications" : [],
            "stack" : [
               {
                  "value" : [
                     {
                        "type" : "Array",
                        "value" : [
                           {
                              "value" : "fZrCUAp6XHd0agwtIy6wf49ymNw=",
                              "type" : "ByteString"
                           },
                           {
                              "value" : "1",
                              "type" : "Integer"
                           },
                           {
                              "type" : "Array",
                              "value" : []
                           },
                           {
                              "type" : "Array",
                              "value" : []
                           },
                           {
                              "type" : "Array",
                              "value" : []
                           }
                        ]
                     }
                  ],
                  "type" : "Array"
               }
            ]
         }
      ],
      "txid" : "0xed1d5081dcef12d2c437ae8566d8040cf0a475e7e354c9acbd0f674a372106f8"
   },
   "id" : 1,
   "jsonrpc" : "2.0"
}
$ curl -sd '{ "jsonrpc": "2.0", "id": 1, "method": "getapplicationlog", "params": ["ed1d5081dcef12d2c437ae8566d8040cf0a475e7e354c9acbd0f674a372106f8"] }' https://rpc.t5.n3.nspcc.ru:20331 | json_pp
{
   "result" : {
      "executions" : [
         {
            "trigger" : "Application",
            "vmstate" : "HALT",
            "gasconsumed" : "98451",
            "stack" : [
               {
                  "value" : [
                     {
                        "type" : "Array",
                        "value" : [
                           {
                              "type" : "ByteString",
                              "value" : "fZrCUAp6XHd0agwtIy6wf49ymNw="
                           },
                           {
                              "value" : "1",
                              "type" : "Integer"
                           },
                           {
                              "value" : [],
                              "type" : "Array"
                           },
                           {
                              "type" : "Array",
                              "value" : []
                           },
                           {
                              "type" : "Array",
                              "value" : []
                           }
                        ]
                     }
                  ],
                  "type" : "Array"
               }
            ],
            "notifications" : [],
            "exception" : null
         }
      ],
      "txid" : "0xed1d5081dcef12d2c437ae8566d8040cf0a475e7e354c9acbd0f674a372106f8"
   },
   "id" : 1,
   "jsonrpc" : "2.0"
}
```
I don't see any FAULTs there. If you can trigger them on-chain --- tell me how. I still think #3034 is a purely CLI issue and it can be fixed without HFs.

Reply:
@roman-khimov @shargon Yes that is correct its purely a CLI issue. Do to `ApplicationEgine` transaction being `null` im guessing. Its only testing the script for invoking methods without transaction attached?

Reply:
Ok! I misunderstand the main issue, I will close it and open a new one